### PR TITLE
Optimize Openstack Swift files download

### LINKED ISSUE #14116: Optimize Openstack Swift large file download
### Steps to reproduce
1. Set Openstack Swift as your primary storage (I'm using one from OVH on S1-8 server).
2. Upload large file (like 512 MB or so).
3. Download it (either desktop client or from web interface).

### Expected behaviour
NextCloud server downloads file segments from object storage and simultaneously transfers them to client.

### Actual behaviour
First entire file is downloaded to NextCloud server (OVH provides ~100Mb download link from their Object Storage from what I can see). Only after it's downloaded it's transferred to client. In the end it takes twice as much time.

I needed to change `max_execution_time` on server to prevent timeout errors.

### Server configuration

Using Docker 15.0.4-php-alpine image.

**Updated from an older Nextcloud/ownCloud or fresh install:** Started at version 14.0

**Signing status:**
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

Enabled:
  - accessibility: 1.1.0
  - activity: 2.8.2
  - admin_audit: 1.5.0
  - calendar: 1.6.4
  - camerarawpreviews: 0.6.3
  - checksum: 0.4.2
  - cloud_federation_api: 0.1.0
  - comments: 1.5.0
  - contacts: 3.0.3
  - data_request: 1.2.0
  - dav: 1.8.1
  - deck: 0.5.2
  - drawio: 0.9.2
  - encryption: 2.3.0
  - event_update_notification: 0.3.2
  - federatedfilesharing: 1.5.0
  - federation: 1.5.0
  - files: 1.10.0
  - files_downloadactivity: 1.4.0
  - files_external: 1.6.0
  - files_markdown: 2.0.5
  - files_mindmap: 0.0.11
  - files_pdfviewer: 1.4.0
  - files_rightclick: 0.11.0
  - files_sharing: 1.7.0
  - files_texteditor: 2.7.0
  - files_trashbin: 1.5.0
  - files_versions: 1.8.0
  - files_videoplayer: 1.4.0
  - firstrunwizard: 2.4.0
  - flowupload: 0.0.9
  - gallery: 18.2.0
  - impersonate: 1.2.0
  - logreader: 2.0.0
  - lookup_server_connector: 1.3.0
  - music: 0.9.3
  - news: 13.0.3
  - nextcloud_announcements: 1.4.0
  - notes: 2.5.1
  - notifications: 2.3.0
  - oauth2: 1.3.0
  - password_policy: 1.5.0
  - previewgenerator: 2.0.0
  - provisioning_api: 1.5.0
  - quota_warning: 1.4.0
  - ransomware_protection: 1.3.0
  - richdocuments: 3.2.0
  - scanner: 0.1.1
  - serverinfo: 1.5.0
  - sharebymail: 1.5.0
  - spreed: 5.0.2
  - support: 1.0.0
  - survey_client: 1.3.0
  - systemtags: 1.5.0
  - tasks: 0.9.8
  - theming: 1.6.0
  - twofactor_backupcodes: 1.4.1
  - twofactor_gateway: 0.11.0
  - twofactor_nextcloud_notification: 1.0.3
  - twofactor_totp: 2.1.0
  - twofactor_u2f: 2.1.0
  - twofactor_yubikey: 0.4.1
  - unsplash: 1.1.3
  - updatenotification: 1.5.0
  - weather: 1.5.5
  - workflowengine: 1.5.0
Disabled:
  - admin_notifications
  - end_to_end_encryption
  - files_accesscontrol
  - files_fulltextsearch
  - fulltextsearch
  - twofactor_rcdevsopenotp
  - user_ldap
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "memcache.local": "\\OC\\Memcache\\APCu",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 6379
        },
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "apps_paths": [
            {
                "path": "\/var\/www\/html\/apps",
                "url": "\/apps",
                "writable": false
            },
            {
                "path": "\/var\/www\/html\/custom_apps",
                "url": "\/custom_apps",
                "writable": true
            }
        ],
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "archive.cloud.pikula.co"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "15.0.4.0",
        "overwrite.cli.url": "https:\/\/new.cloud.pikula.co",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "objectstore": {
            "class": "OC\\Files\\ObjectStore\\Swift",
            "arguments": {
                "username": "dEcCjjVA5ghy",
                "password": "***REMOVED SENSITIVE VALUE***",
                "bucket": "cloud.pikula.co-data",
                "autocreate": false,
                "region": "WAW1",
                "url": "https:\/\/auth.cloud.ovh.net\/v2.0",
                "tenantName": "7435064151262691",
                "serviceName": "swift"
            }
        },
        "maintenance": false,
        "loglevel": 2,
        "theme": ""
    }
}
```
</details>

**Are you using external storage, if yes which one:** OpenStack Swift as primary storage

**Are you using encryption:** no

**Are you using an external user-backend, if yes which one:** No

### Client configuration
**Browser:** Firefox 65.0 and NextCloud client 2.5.1final (build 20181204) AppImage

**Operating system:** 
<details>
```
NAME=Fedora
VERSION="29 (Server Edition)"
ID=fedora
VERSION_ID=29
VERSION_CODENAME=""
PLATFORM_ID="platform:f29"
PRETTY_NAME="Fedora 29 (Server Edition)"
ANSI_COLOR="0;34"
LOGO=fedora-logo-icon
CPE_NAME="cpe:/o:fedoraproject:fedora:29"
HOME_URL="https://fedoraproject.org/"
DOCUMENTATION_URL="https://docs.fedoraproject.org/en-US/fedora/f29/system-administrators-guide/"
SUPPORT_URL="https://fedoraproject.org/wiki/Communicating_and_getting_help"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_BUGZILLA_PRODUCT="Fedora"
REDHAT_BUGZILLA_PRODUCT_VERSION=29
REDHAT_SUPPORT_PRODUCT="Fedora"
REDHAT_SUPPORT_PRODUCT_VERSION=29
PRIVACY_POLICY_URL="https://fedoraproject.org/wiki/Legal:PrivacyPolicy"
VARIANT="Server Edition"
VARIANT_ID=server
```
</details>

### Logs
Well, in the end it works, but not in the optimal way. No errors are present (besides timeout, which was mitigated by changing timeout in php.ini).

### PR DESCRIPTION
**Pull request by adrb**
fix #14116

It improves efficiency when downloading files from Swift storage.
Before, files were downloaded and then pushed back to user.
That behaevior causes all kinds of performance problems.

Now, files are streamed directly to user.

### CODE REVIEW

**Comment by adrb** [comment]
I see that some tests failed but it doesn't seems that it was caused by patch, for e.g. [that one](https://drone.nextcloud.com/nextcloud/server/25133/16/5)

**Comment by kesselb** [comment]
How does this optimization work? ;) I don't have a setup with openstack for testing. Out of curiosity: What happens with the patch below?

```
Index: lib/private/Files/ObjectStore/Swift.php
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- lib/private/Files/ObjectStore/Swift.php	(revision 092a1fb90f57639a15adc5a8e565b53debe26a8a)
+++ lib/private/Files/ObjectStore/Swift.php	(date 1579009971686)
@@ -107,12 +107,12 @@
 			}
 		}
 		$objectContent->rewind();
+//
+//		$stream = $objectContent->detach();
+//		// save the object content in the context of the stream to prevent it being gc'd until the stream is closed
+//		stream_context_set_option($stream, 'swift', 'content', $objectContent);
 
-		$stream = $objectContent->detach();
-		// save the object content in the context of the stream to prevent it being gc'd until the stream is closed
-		stream_context_set_option($stream, 'swift', 'content', $objectContent);
-
-		return RetryWrapper::wrap($stream);
+		return RetryWrapper::wrap($objectContent);
 	}
 
 	/**
```

**Comment by adrb** [comment]
Now it works pretty much the same as for S3. It setup authorization headers and then stream Swift file with fopen.

I don't think that patch you quoted will cause any difference since it will still waits for download to complete before returning from readObject function

**Comment by leo-lb** [comment]
What about upload as well?

**Comment by icewind1991** [comment]
The old code should already do streaming downloads from how I understand the openstack sdk/guzzle

**Comment by leo-lb** [comment]
> The old code should already do streaming downloads from how I understand the openstack sdk/guzzle

I could experience that it did not. I had to wait for the file to be downloaded fully on the server to be served a download in my browser.

**Comment by adrb** [comment]
> What about upload as well?

I didn't touch that yet. But plan is to migrate my installation to Swift and upload also needs to handle large objects. However I already saw that at moment it don't, so I will probably take care of it too

> > The old code should already do streaming downloads from how I understand the openstack sdk/guzzle
> 
> I could experience that it did not. I had to wait for the file to be downloaded fully on the server to be served a download in my browser.

That's correct, It streams **already downloaded** files, which results in excessive disk space and I/O consumption. On top of that, waiting for multi gigabyte file results in connection timeout which is devastating, because what users do next? Hit the refresh button, and keep in mind that earlier file is still being downloaded ;)

**Comment by maniac777** [comment]
@kesselb regarding your patch it does not change nature of the issue as file is downloaded to nextcloud few lies earlier.

```	
public function readObject($urn) {
		try {
			$object = $this->getContainer()->getObject($urn);
			// we need to keep a reference to objectContent or
			// the stream will be closed before we can do anything with it
			$objectContent = $object->download();
```
here exacly to be precise - in download().

@icewind1991: unfortunetly the old code calls download() function from php-opencloud library. Just for conveniece I'm pasting it below. I believe that file is downloaded due to calling execute(). It blocks and saves entire file in temp dir.

```
    /**
     * This call will perform a `GET` HTTP request for the given object and return back its content in the form of a
     * Guzzle Stream object. Downloading an object will transfer all of the content for an object, and is therefore
     * distinct from fetching its metadata (a `HEAD` request). The body of an object is not fetched by default to
     * improve performance when handling large objects.
     *
     * @param array $data {@see \OpenStack\ObjectStore\v1\Api::getObject}
     */
    public function download(array $data = []): StreamInterface
    {
        $data += ['name' => $this->name, 'containerName' => $this->containerName];
        /** @var ResponseInterface $response */
        $response = $this->execute($this->api->getObject(), $data);
        $this->populateHeaders($response);
        return $response->getBody();
    }

```

**Comment by kesselb** [comment]
> regarding your patch it does not change nature of the issue as file is downloaded few lines of code earlier.

OK ;) I think we should keep the `BadResponseError` handling. That can still happen. You probably have to call `$object->retrieve` to trigger it.

**Comment by adrb** [comment]
> OK ;) I think we should keep the `BadResponseError` handling. That can still happen. You probably have to call `$object->retrieve` to trigger it.

Even after successful HEAD request from `$object->retrieve`, actual reading with `fopen` can fail, so i'm not sure if it is worth it

**Comment by icewind1991** [comment]
Ok, fair enough, I'm ok with this approach, thanks for the thorough explanation.

you should be able to remove the call to `getObject` I think to reduce overhead, since publicUri can be constructed from the bucket and object name.

Also a line or 2 documentation for `getHeaders` would be appreciated :)

**Comment by kesselb** [comment]
> Even after successful HEAD request from $object->retrieve, actual reading with fopen can fail, so i'm not sure if it is worth it

https://github.com/nextcloud/server/blob/5bf3d1bb384da56adbf205752be8f840aac3b0c5/lib/private/Files/ObjectStore/ObjectStoreStorage.php#L288-L302

A `NotFoundException` is handled different. https://github.com/nextcloud/server/pull/12500 introduced it. Probably @icewind1991 is able to tell if we still need to handle 404 different.

**Comment by icewind1991** [comment]
For error handling you'll need to check the result from fopen and use [$http_response_header](https://www.php.net/manual/en/reserved.variables.httpresponseheader.php) to throw the relevant exceptions or return false.

**Comment by icewind1991** [comment]
Since the `cachedToken` is not always set when constructing the swift client, some extra logic is needed to set it after authenticating, something like 

https://gist.github.com/icewind1991/3f3088c17917c783ceb21d002e966727

**Comment by kesselb** [comment]
Thanks @icewind1991 :+1: 

I would also prefer to not add the `getHeaders` method to `SwiftFactory`. Of course it aligns with `S3ObjectTrait` but it's misleading IMO. A method to fetch the cached token from `SwiftFactory` should be enough. We can still add a `getHeaders` method later if needed.

**Comment by adrb** [comment]
Thank you for your input, pushed second version, I hope that meets your expectations.

@icewind1991 
> you should be able to remove the call to `getObject` I think to reduce overhead, since publicUri can be constructed from the bucket and object name.

Object store url is not the same as in `swiftFactory::params['url']`, and I didn't found simple solution for doing that. Second thing, it can be changed in further Swift versions. 

@kesselb
> I would also prefer to not add the `getHeaders` method to `SwiftFactory`. Of course it aligns with `S3ObjectTrait` but it's misleading IMO. A method to fetch the cached token from `SwiftFactory` should be enough. We can still add a `getHeaders` method later if needed.

Since constructing `$cacheKey` is not easy thing, we need some kind of helper method. But I agree that `getHeaders` wasn't the best solution.

File: lib/private/Files/ObjectStore/Swift.php [hunk 1]
```diff
@@ -89,30 +95,33 @@ public function writeObject($urn, $stream) {
...
 			}
 		}
-		$objectContent->rewind();
-
-		$stream = $objectContent->detach();
-		// save the object content in the context of the stream to prevent it being gc'd until the stream is closed
-		stream_context_set_option($stream, 'swift', 'content', $objectContent);
 
-		return RetryWrapper::wrap($stream);
+		return $response->getBody()->detach();
```
Comment (icewind1991) [inline lib/private/Files/ObjectStore/Swift.php]: We should probably keep the retry wrapper on

**Comment by icewind1991** [comment]
Looks good otherwise

thanks for the work :+1:

File: lib/private/Files/ObjectStore/Swift.php [hunk 2]
```diff
@@ -89,30 +95,33 @@ public function writeObject($urn, $stream) {
...
+					'stream' => true,
+					'headers' => [
+						'X-Auth-Token' => $tokenId,
+						'Cache-Control' => 'no-cache'
+					],
+				]
+			);
+
+		} catch (BadResponseException $e) {
 			if ($e->getResponse()->getStatusCode() === 404) {
```
Comment (kesselb) [inline lib/private/Files/ObjectStore/Swift.php]: Please check `getResponse` for null.

**Review by kesselb — COMMENTED** [review]
Thanks :+1: Great contribution :tada:

**Review by icewind1991 — APPROVED** [review]

**Comment by skjnldsv** [comment]
Ping! :bell:

**Review by rullzer — APPROVED** [review]
:elephant:

**Comment by rullzer** [comment]
@adrb could you rebase this on latest master so the full CI can run one more time and then we get it in

**Comment by adrb** [comment]
Could you also review changes from #18955? It's complementary to this one

**Comment by kesselb** [comment]
Thanks @adrb 👍

**Pull request merged**